python/lsst/afw/image/multiband.py

```diff
@@ -0,0 +1,1093 @@
+        """
+        self._image = MultibandImage(filters, image)
+        self._mask = MultibandMask(filters, mask)
+        self._variance = MultibandImage(filters, variance)
+
+    def setBBox(self, bbox):
+        """Set the bounding box
+
+        This is different than `MultibandBase.setBBox`
+        because the multiband `image`, `mask`, and `variance` objects
+        must all have their bounding boxes updated.
+
+        Parameters
+        ----------
+        bbox: `Box2I` or tuple of indices
+            Bounding box to set as the current bounding box.
+            If a tuple of indices is passed, a new bounding box
+            is created from them and used.
+        """
+        super().setBBox(bbox)
+        self.image.setBBox(bbox)
+        self.mask.setBBox(bbox)
+        self.variance.setBBox(bbox)
+
+    def copy(self, deep=False):
+        """Make a copy of the current instance
+        """
+        if deep:
+            filters = tuple([f for f in self.filters])
+            singles = tuple([self.singleType(s, deep=True) for s in self.singles])
+            result = type(self)(filters=filters, singles=singles)
+        else:
+            result = type(self)(image=self.image, mask=self.mask, variance=self.variance,
+                                filters=self.filters)
+        return result
+
+    def _slice(self, filters, filterIndex, indices):
+        """Slice the current object and return the result
+
+        See `Multiband._slice` for a list of the parameters.
+        """
+        image = self.image._slice(filters, filterIndex, indices)
+        mask = self._mask._slice(filters, filterIndex, indices)
+        variance = self._variance._slice(filters, filterIndex, indices)
+
+        result = type(self)(filters=filters, image=image, mask=mask, variance=variance)
+        assert np.all([r.getBBox() == result._bbox for r in [result._mask, result._variance]])
+        return result
+
+    def _verifyUpdate(self, image=None, mask=None, variance=None):
+        """Check that the new image, mask, or variance is valid
+
+        This basically means checking that the update to the
+        property matches the current bounding box and inherits
+        from the `MultibandBase` class.
+        """
+        for prop in [image, mask, variance]:
+            if prop is not None:
+                if prop.getBBox() != self.getBBox():
+                    raise ValueError("Bounding box does not match the current class")
+                if not isinstance(prop, MultibandBase):
+                    err = "image, mask, and variance should all inherit from the MultibandBase class"
+                    raise ValueError(err)
+
+    def getImage(self):
+        """Get the image
+        """
+        return self._image
+
+    def setImage(self, image):
+        """Set the image
+        """
+        self._verifyUpdate(image=image)
+        self._image = image
+        self._singles = self._buildSingles(image=image)
```

Comment:
I included the setters so that `image`, `mask`, and `variance` can be properties (similar to `Exposure`).

Reply:
I like the idea of making them properties, but the question of what a setter should do here (copy pixel values?  reset the "pointer" to the given object, even if that messes up the 3-d data cube?) is sufficiently non-obvious (and the need for a setter is sufficiently unproven) that I'm not sure we should allow them to be changed.

Reply:
I don't follow you, or see how this is different than the `MaskedImage` usage:
```
// in pybind11 wrapper:
    cls.def("getImage", &MI::getImage);
    cls.def("setImage", &MI::setImage);
    cls.def_property("image", &MI::getImage, &MI::setImage);
```
```
// C++ header:
void setImage(Image const& other) { _image->assign(other); }
```

Our coding guidelines require properties to have setters, right? So due to our guidelines we need to have setters to be able to use the (very convenient) `image`, `mask`, and `variance` properties.

Maybe it isn't clear how the data is stored. The `MultibandTriple.image` is a `MultibandImage`, `MultibandTriple.mask` is a `MultibandMask`, and `MultibandTripleBase.variance` is a `MultibandImage`, each of which has it's own internal 3D data cube. The `MultibandTripleBase` is just a container to access the image, mask, and variance, not a 4D data cube that combines them all, so reassigning `MultibandTriple.image` just assigns the new `MultibandImage` to the internal property, which seems to be the desired behavior. Does that make sense?

Reply:
Our coding guidelines do not require properties to have setters.  But of course you're completely right that the setter here isn't problematic, because you're updating the entire data cube at a time.  Sorry for the noise.

Reply:
No worries. After responding I noticed your comment that the docs need to be improved to explain how the data is stored in `MultibandMaskedImage`, so I can see how this might be confusing.

But just to clarify, I thought that properties that point to mutable objects are required to have setters to avoid unexpected behavior (which is why we couldn't have a `bbox` property). Is there a subtlety that I'm missing?

Reply:
The requirement is that properties that point to mutable objects can't return copies, so you can't call `a.b.mutate()` without modifying `a.b`.

Requiring or at least strongly encouraging such properties to also have setters would be a good idea too, but I think there may be some cases where the setter would always be a bit confusing (just because `a.b = x` can do a deep-copy of some kind while `b = a.b; b = x` cannot), so I think encouraging the existence of a setter may be all we should do.

Reply:
Oh, I see. I never looked at `Image` internally and didn't realize that `bbox` is not stored anywhere but constructed from `_origin` and the dimensions. This makes sense now.